### PR TITLE
More granular control on checks made by CLI commands

### PR DESCRIPTION
**Pull request by alpeb**
Ref #1854

Let me know what you think of this approach.
As a preliminary test, with this change, after deleting the Grafana deployment, `linkerd stat` works fine and `linkerd dashboard` errors out.

### CODE REVIEW

File: pkg/healthcheck/healthcheck.go [hunk 1]
```diff
@@ -749,6 +773,14 @@ func validateControlPlanePods(pods []v1.Pod) error {
 	return nil
 }
 
+func checkControllerRunning(pods []v1.Pod) error {
+	statuses := getPodStatuses(pods)
+	if _, ok := statuses["controller"]; !ok {
+		return fmt.Errorf("No running pods for \"linkerd-controller\"")
```
[the remark this answers is not in the document]
Comment (alpeb) [reply pkg/healthcheck/healthcheck.go]: Indeed!

**Comment by alpeb** [comment]
@siggy Ok cool. I wasn't aware until now that `options` stemmed really from `check` command-line options, and it makes sense to use them just for that.

On the other hand, I had added a check against  `version.CheckServerVersion()` for querying the control plane's Version endpoint, but you're instead using `version.GetLatestVersion()` which queries our servers for the latest version. Is that intended?

**Comment by siggy** [comment]
@alpeb Ah nice catch! You're correct that the cli commands should be calling some variant of `version.CheckServerVersion()`. What do you think about just hard-coding `ShouldCheckControlPlaneVersion: true` in `validatedPublicAPIClient`?

For `version.GetLatestVersion()`, let's just include it anyway, with the modification to make it a warning rather than fatal.

**Comment by alpeb** [comment]
@siggy Sounds good to me :+1:

File: cli/cmd/get.go [hunk 2]
```diff
@@ -56,7 +55,7 @@ Only pod resources (aka pods, po) are supported.`,
 				return fmt.Errorf("invalid resource type %s, valid types: %s", friendlyName, k8s.Pod)
 			}
 
-			podNames, err := getPods(validatedPublicAPIClient(time.Time{}), options)
+			podNames, err := getPods(cliPublicAPIClient(), options)
```
[the remark this answers is not in the document]
Comment (siggy) [reply cli/cmd/get.go]: @dadjeibaah I had suggested the addition of `cliPublicAPIClient()`, as there are 5 call sites that do `validatedPublicAPIClient(time.Time{}, false)`, while only `dashboard.go` needs to call `validatedPublicAPIClient` with non-default params.

File: pkg/version/version.go [hunk 3]
```diff
@@ -45,6 +45,7 @@ func CheckClientVersion(expectedVersion string) error {
 	return nil
 }
 
+//
```
[the remark this answers is not in the document]
Comment (alpeb) [reply pkg/version/version.go]: Looks like this refers to the version I had before I force-pushed adding @siggy suggested changes and removing mine. Am I breaking etiquette by doing force-pushes into PRs?

File: pkg/healthcheck/healthcheck.go [hunk 4]
```diff
@@ -316,8 +328,8 @@ func (hc *HealthChecker) addLinkerdAPIChecks() {
 	})
 
 	hc.checkers = append(hc.checkers, &checker{
-		category:      LinkerdAPICategory,
-		description:   "control plane pods are ready",
+		category:      LinkerdExistenceCategory,
+		description:   "controler pod is running",
```
[the remark this answers is not in the document]
Comment (alpeb) [reply pkg/healthcheck/healthcheck.go]: Nice catch! :mag:

File: pkg/version/version.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (siggy) [reply pkg/version/version.go]: @alpeb No worries! In general we try to avoid force-pushes in PR, to show reviewers the incremental changes as we go along.

That said, I think this case is an exception where the initial PR was branched off an older master, already had conflicts, and the subsequent commit was quite different from the initial one. I probably would have force-pushed in this case.

File: pkg/healthcheck/healthcheck.go [hunk 5]
```diff
@@ -15,7 +16,7 @@ import (
 	"github.com/linkerd/linkerd2/pkg/profiles"
 	"github.com/linkerd/linkerd2/pkg/version"
 	authorizationapi "k8s.io/api/authorization/v1beta1"
-	"k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
```
[the remark this answers is not in the document]
Comment (alpeb) [reply pkg/healthcheck/healthcheck.go]: Looks like a bug in VSCode that insists on adding that. I'll address.

File: pkg/healthcheck/healthcheck.go [hunk 6]
```diff
@@ -44,8 +45,14 @@ const (
 	// checks must be added first.
 	LinkerdDataPlaneChecks
 
+	// LinkerdExistenceChecks adds a series of checks to validate that the control
+	// plane namespace and controller pod exist.
+	// These checks are dependent on the output of KubernetesAPIChecks, so those
+	// checks must be added first.
+	LinkerdExistenceChecks
```
[the remark this answers is not in the document]
Comment (alpeb) [reply pkg/healthcheck/healthcheck.go]: `LinkerdControlPlaneExistenceChecks` is fine by me :+1:

File: pkg/healthcheck/healthcheck.go [hunk 7]
```diff
@@ -414,7 +443,7 @@ func (hc *HealthChecker) addLinkerdVersionChecks() {
 	hc.checkers = append(hc.checkers, &checker{
 		category:    LinkerdVersionCategory,
 		description: "can determine the latest version",
-		fatal:       true,
+		warning:     true,
```
Comment (klingerf) [inline pkg/healthcheck/healthcheck.go]: This check needs to remain as fatal, unfortunately, since knowing the latest version is a prerequisite for all of the subsequent version checks. If we can't reach versioncheck.linkerd.io to determine the latest version, then we can't validate that any of the components are running the latest version or not.

Also, since the "control plane is up-to-date" check is already marked as a warning, if that check fails because the `/Version` endpoint is unreachable, that won't be reported back to the CLI. Instead, the CLI will ignore the failure, and return the client, even though the client (assumedly) won't be able to connect to the public API.

I recommend reverting this change and adding a separate set of checks, similar to `LinkerdAPIChecks`, that calls the `/Version` endpoint instead of the `/SelfCheck` endpoint. Then we can continue use `LinkerdAPIChecks` for the `linkerd check` and `linkerd dashboard` commands, and we can use the new set of more limited checks (control plane pod exists and `/Version` endpoint returns without error) for all other commands.

File: cli/cmd/root.go [hunk 8]
```diff
@@ -85,22 +85,36 @@ func init() {
...
-		KubeConfig:            kubeconfigPath,
-		KubeContext:           kubeContext,
-		APIAddr:               apiAddr,
-		RetryDeadline:         retryDeadline,
+		ControlPlaneNamespace:          controlPlaneNamespace,
+		KubeConfig:                     kubeconfigPath,
+		KubeContext:                    kubeContext,
+		APIAddr:                        apiAddr,
+		RetryDeadline:                  retryDeadline,
+		ShouldCheckControlPlaneVersion: true,
```
Comment (klingerf) [inline cli/cmd/root.go]: Related to my comment below, let's revert this change.

File: pkg/healthcheck/healthcheck.go (hunk 7 above)
Comment (alpeb) [reply pkg/healthcheck/healthcheck.go]: @klingerf I agree with the reasoning for leaving the version check as a fatal check.

But I think what you're asking for in your last paragraph has already been addressed.
With the last commit we extracted out of `LinkerdAPIChecks` the checks for (using same wording as #1854):
- control plane namespace exists
- can initialize the control plane API client

and added a new check for:
- "linkerd-controller-*" pod is running (not necessarily ready)

Those were put into the new `LinkerdExistenceChecks` (which doesn't include `/SelfCheck`) that are used by all the CLI commands but `linkerd check` and `linkerd dashboard`. 

`linkerd check` and `linkerd dashboard` continue making the same set of strict checks since they use both `LinkerdExistenceChecks` and `LinkerdAPIChecks`.

Another change was that, before, only `linkerd check` was using `LinkerdVersionChecks` (containing `/Version` and the other related checks). Now all the CLI commands use that as well.

@grampelberg might wanna want to chime in since these changes were really his :slightly_smiling_face:

Comment (siggy) [reply pkg/healthcheck/healthcheck.go]: @klingerf Yup, that fatal -> warning change was mine. The intent was to allow cli commands to run if versioncheck.linkerd.io fails, but you're correct that we should just not run the check, and for the `linkerd check` and `dashboard` cases it should fatally fail the subsequent version checks.

@alpeb's correct that we've already broken out a limited set of checks for all commands except `dashboard` and `check`, specifically:
- `KubernetesAPIChecks`
- `LinkerdExistenceChecks`
- `LinkerdVersionChecks`

It sounds like we should not perform `LinkerdVersionChecks` for any commands except `dashboard` and `check`. We can do that by modifying `validatedPublicAPIClient` as follows:

```go
checks := []healthcheck.Checks{
	healthcheck.KubernetesAPIChecks,
	healthcheck.LinkerdExistenceChecks,
}

if apiChecks {
	checks = append(checks, healthcheck.LinkerdAPIChecks)
	checks = append(checks, healthcheck.LinkerdVersionChecks)
}
```

...and then introduce a new type of `/Version` check (which I think you had in your initial PR), and include that as part of `LinkerdExistenceChecks`, something like:

```go
func GetServerVersion(apiClient pb.ApiClient) (string, error) {
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()

	rsp, err := apiClient.Version(ctx, &pb.Empty{})
	if err != nil {
		return nil, err
	}

	return  rsp.GetReleaseVersion(), nil
}
```

... and then we could refactor `CheckServerVersion` to use this new `GetServerVersion` function.

For reference, here's the current `linkerd check` output for this branch:
```
kubernetes-api: can initialize the client..................................[ok]
kubernetes-api: can query the Kubernetes API...............................[ok]
kubernetes-api: is running the minimum Kubernetes API version..............[ok]
linkerd-existence: control plane namespace exists..........................[ok]
linkerd-existence: controler pod is running................................[ok]
linkerd-existence: can initialize the client...............................[ok]
linkerd-api: control plane pods are ready..................................[ok]
linkerd-api: can query the control plane API...............................[ok]
linkerd-api[kubernetes]: control plane can talk to Kubernetes..............[ok]
linkerd-api[prometheus]: control plane can talk to Prometheus..............[ok]
linkerd-api: no invalid service profiles...................................[ok]
linkerd-version: can determine the latest version..........................[warning] -- Unsupported version channel: git
linkerd-version: cli is up-to-date.........................................[warning] -- is running version 0c7049f0 but the latest version is
linkerd-version: control plane is up-to-date...............................[warning] -- is running version 18.12.4 but the latest version is
```

Comment (klingerf) [reply pkg/healthcheck/healthcheck.go]: Ok great, thanks for the additional info. The updated approach that @siggy laid out makes sense to me.

I'm sorry the code is so convoluted. To help with my own understanding, I made a list of all of the checks that we run when building the public API client that we use when executing the `linkerd stat` command (for example). As this branch is currently written, they are as follows:

* kubernetes-api: can initialize the client
* kubernetes-api: can query the Kubernetes API
* linkerd-existence: control plane namespace exists
* linkerd-existence: controller pod is running
* linkerd-existence: can initialize the client
* linkerd-version: can determine the latest version
* linkerd-version: cli is up-to-date
* linkerd-version: control plane is up-to-date

Those last 3 "linkerd-version" checks will never result in an error message, however, since they are all marked as warnings. The [exitOnError func](https://github.com/linkerd/linkerd2/blob/master/cli/cmd/root.go#L106) that we use when running the checks explicitly ignores all failed checks that are marked as warnings, so they're effectively a no-op.

In order to bring those checks in line with what's described in https://github.com/linkerd/linkerd2/issues/1854#issuecomment-447976710, we should change them as follows:

* kubernetes-api: can initialize the client
* kubernetes-api: can query the Kubernetes API
* linkerd-existence: control plane namespace exists
* linkerd-existence: controller pod is running
* linkerd-existence: can initialize the client
* __linkerd-existence: can successfully call the `/Version` endpoint__
* ~~linkerd-version: can determine the latest version~~
* ~~linkerd-version: cli is up-to-date~~
* ~~linkerd-version: control plane is up-to-date~~

And I agree that updating `addLinkerdExistenceChecks` to include that one additional check and removing the call to `addLinkerdVersionChecks` should do the trick.

**Comment by alpeb** [comment]
@klingerf @siggy turns out making that version check fatal breaks using `linkerd dashboard` locally (and thus the integration tests as well):
```
linkerd-version: can determine the latest version..........................[FAIL] -- Unsupported version channel: dev
```
What do you think about adding a condition in `GetLatestVersion` to have it not fail in case of `Version == "dev"` ?

**Comment by klingerf** [comment]
I think we should go back to only running the version checks as part of the `linkerd check` command. There's no need to check the version when running `linkerd dashboard`, since the dashboard performs a version check once it loads.

I'd definitely be in favor of improving the handling of the "dev" and "git" version prefixes, but we should do that in a separate branch.

File: pkg/healthcheck/healthcheck.go [hunk 9]
```diff
@@ -344,6 +356,33 @@ func (hc *HealthChecker) addLinkerdAPIChecks() {
 		},
 	})
 
+	hc.checkers = append(hc.checkers, &checker{
+		category:    LinkerdControlPlaneExistenceCategory,
+		description: "can successfully call the /Version endpoint",
+		fatal:       true,
```
Comment (siggy) [inline pkg/healthcheck/healthcheck.go]: i think this needs `retryDeadline: hc.RetryDeadline`, otherwise `linkerd check` will fail immediately and exit without retry.

you can repro this by running `linkerd install` and then quickly running `linkerd check`...

```bash
$ bin/linkerd check
kubernetes-api: can initialize the client..................................[ok]
kubernetes-api: can query the Kubernetes API...............................[ok]
kubernetes-api: is running the minimum Kubernetes API version..............[ok]
linkerd-existence: control plane namespace exists..........................[ok]
linkerd-existence: controller pod is running...............................[ok]
linkerd-existence: can initialize the client...............................[ok]
linkerd-existence: can successfully call the /Version endpoint.............[FAIL] -- Unexpected API response: 503 Service Unavailable

Status check results are [FAIL]
```

File: pkg/healthcheck/healthcheck.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (siggy) [reply pkg/healthcheck/healthcheck.go]: bump ;)

**Comment by klingerf** [comment]
(you can ignore my comment about making the check non-fatal; that was leftover from a previous revision and i deleted it)

File: cli/cmd/root.go [hunk 10]
```diff
@@ -114,8 +125,12 @@ func validatedPublicAPIClient(retryDeadline time.Time) pb.ApiClient {
 			switch result.Category {
 			case healthcheck.KubernetesAPICategory:
 				msg = "Cannot connect to Kubernetes"
+			case healthcheck.LinkerdControlPlaneExistenceCategory:
+				msg = "Cannot find Linkerd"
 			case healthcheck.LinkerdAPICategory:
 				msg = "Cannot connect to Linkerd"
+			case healthcheck.LinkerdVersionCategory:
+				msg = "Invalid Linkerd version"
```
[the remark this answers is not in the document]
Comment (alpeb) [reply cli/cmd/root.go]: Good catch :+1:

File: pkg/healthcheck/healthcheck.go [hunk 11]
```diff
@@ -344,6 +356,34 @@ func (hc *HealthChecker) addLinkerdAPIChecks() {
 		},
 	})
 
+	hc.checkers = append(hc.checkers, &checker{
+		category:      LinkerdControlPlaneExistenceCategory,
+		description:   "can successfully call the /Version endpoint",
```
[the remark this answers is not in the document]
Comment (alpeb) [reply pkg/healthcheck/healthcheck.go]: Sounds good to me :+1:
